fix: toggling initialized values from multi selections

arvind opened the pull request:
Fixes #5199, which revealed several minor bugs:

1. Now that initial values for selections are embedded directly as dataset values (rather than embedded in signal expressions), we no longer need to string escape model names. 

2. Datetime expressions should evaluate to numerical timestamps, rather than strings. 

3. Vega's `modify` expression should also test for object equality (which wasn't previously necessary for our logic because the same object was being reused in every tuple and thus strict equality testing `===` would correctly evaluate to `true`). vega/vega#1909.

4. ~~Although the typings/schema disallow it (and I believe should continue to disallow it, for simplicity/disambiguity), when a multi selection is projected over a single field, it is reasonable to expect people to use a short form initialization as seen in #5199. i.e., `{init: {Origin: ['Japan', 'USA']}}`. Now, we test for this one corner case, and expand it to what it should be `{init: [{Origin: 'Japan'}, {Origin: 'USA'}]}`~~

arvind commented:
Marking this as a WIP -- after reading #4766, I want to consider whether there's a straightforward way of supporting columnar initialization of multi-selections which is probably the more correct fix than the awkward workaround this PR currently proposes. 

What makes columnar support slightly tricky is that to initialize ranged fields (e.g., binned fields), users specify an array of values. Thus, we need to distinguish `"init": {"x": [0, 100], "Origin": "Europe"}` (which should initialize just a single tuple because `x` is a ranged field) from `"init": {"x": [0, 100], "Origin": ["Europe", "Japan"]}` which should generate two tuples `(0, Europe), (100, Japan)`. Perhaps we test to ensure all arrayed properties have the same length?

domoritz commented:
What if x and y are ranged fields? `"init": {"x": [0,20], "y": [10,30]}`? Does this mean `[[0,10],[20,30]]` or `[[0,20],[10,30]]`?

gordonwoodhull commented:
Aha! Makes sense - confirmed, it does work if I use long form.

I guess I got the short form from the selection event, tried it, it seemed to work except for this tiny glitch, didn't think I might have the wrong format.

I validate with ajv every time I render, but I find the errors difficult to interpret. I have a few cases where things seem to work but validation fails. At some point I'll take the time to figure out whether vega-lite is being more permissive than the schema, or if I'm just passing extra/irrelevant/incomplete bits of spec due to generation.

arvind commented:
> What if x and y are ranged fields?

If both `x` and `y` are ranged fields, then that `init` value would just be transformed trivially to `"init": [{"x": [0,20], "y": [10,30]}]` and stored as `"fields":  ["x", "y"], "values": [[0, 20], [10, 30]]"`.

arvind commented:
I've removed the `WIP` label and dropped the commit related to columnar/shortform initialization of multi selections (ae7f0d4ae88593a35bd7bb9b4c11194564052f70) to keep this PR focused on bug fixes and unblock it for review. 

Columnar initialization is proving to be slightly trickier than I anticipated, and I exceeded my time box for it (it doesn't seem to be a sufficiently high priority for now given we support long form initialization). We're tracking this as an enhancement in #4766.

arvind commented:
@gordonwoodhull thanks for confirming things work with long form! Agreed on the fact that the schema can sometimes throw confusing errors, and I think this is especially true for how selections can be initialized. I think this is where we can do better with additional documentation.

domoritz commented on src/compile/selection/index.ts:
```diff
@@ -87,8 +87,8 @@ function getFacetModel(model: Model): FacetModel {
 
-export function unitName(model: Model) {
-  let name = stringValue(model.name);
+export function unitName(model: Model, escape = true) {
```
We prefer to use objects for boolean properties to make them more understandable. 

```suggestion
export function unitName(model: Model, {escape} = {escape: true}) {
```

arvind replied:
Thanks for the suggestion, @domoritz! Implemented in 283419f.
